packages/pool/src/sim/index_out.rs

```diff
@@ -0,0 +1,184 @@
+            .map(|v| (v.get_id(), v.clone()))
+            .collect::<BTreeMap<_, _>>();
+
+        let routes_with_amount = self
+            .index_units
+            .iter()
+            .map(|(denom, unit)| coin((token_out * *unit).u128(), denom))
+            .map(|token_out| {
+                // if this unit does not have to do swap
+                if token_out.denom == input_asset {
+                    Ok(SimAmountInRoute {
+                        // amount_in = amount_out
+                        sim_amount_in: token_out.amount,
+                        amount_out: token_out,
+                        routes: None,
+                    })
+                } else {
+                    self.estimate_in_given_out(input_asset, token_out, &mut pools_map)
+                }
+            })
+            .collect::<Result<Vec<_>, PoolError>>()?;
+
+        let total_spent = routes_with_amount
+            .iter()
+            .try_fold(Uint128::zero(), |acc, v| {
+                Ok::<_, PoolError>(acc.checked_add(v.sim_amount_in)?)
+            })?;
+
+        let ret = SimIndexOutResp {
+            total_spent,
+            sim_routes: SimAmountInRoutes(routes_with_amount),
+        };
+
+        Ok(ret)
+    }
+
+    fn search_efficient_amount_for_input_f(
+        &self,
+        desired_input: Coin,
+        token_test_out: Uint128,
+        token_test_in: Uint128,
+    ) -> Result<SearchAmountForInputResp, PoolError> {
+        let est_min_token_out =
+            token_test_out * Decimal::checked_from_ratio(desired_input.amount, token_test_in)?;
+
+        let res = self.estimate_token_given_index_out(est_min_token_out, &desired_input.denom)?;
+
+        let ret = SearchAmountForInputResp {
+            max_token_in: res.total_spent,
+            est_min_token_out,
+            sim_routes: res.sim_routes,
+        };
+
+        Ok(ret)
+    }
+
+    pub fn search_efficient_amount_for_input(
+        &self,
+        desired_input: Coin,
+        init_output_amount: Option<Uint128>,
+    ) -> Result<SearchAmountForInputResp, PoolError> {
+        let mut acc_res = {
+            let token_out = init_output_amount.unwrap_or(Uint128::new(1000000));
+            let est_res = self.estimate_token_given_index_out(token_out, &desired_input.denom)?;
+
+            SearchAmountForInputResp {
+                max_token_in: est_res.total_spent,
+                est_min_token_out: token_out,
+                sim_routes: est_res.sim_routes,
+            }
+        };
+
+        let mut loop_count = 0;
+        loop {
+            if loop_count >= MAX_LOOP {
+                return Err(PoolError::MaxLoopExceeded);
+            }
+
+            acc_res = self.search_efficient_amount_for_input_f(
+                desired_input.clone(),
+                acc_res.est_min_token_out,
+                acc_res.max_token_in,
+            )?;
+
+            let err = Decimal::checked_from_ratio(desired_input.amount, acc_res.max_token_in)?;
```

Comment:
It's so strange that the err variable isn't the error type.

Reply:
oh it's literally mathematical error

Reply:
how could `<=` operator applied for Error type? I think it is not Error type because `err` applied `<=` operator below this code @byeongsu-hong

Reply:
no it's literally mathematical error